File: lib/core/utils/bold_markup.dart
/// A minimal inline-markup parser for Knowledge Note bodies.
///
/// The grammar is deliberately tiny — NOT Markdown:
/// * `\n\n` (one or more blank lines) splits paragraphs; each paragraph is
///   trimmed and empty paragraphs are dropped.
/// * `**x**` (x non-empty, no nested `**`, non-greedy) marks a bold run.
/// * Everything else is literal — a lone `**` renders as the two characters,
///   and `*` / `_` / `#` are plain text. No nesting, italics, or links.
library;

/// One styled span of text within a paragraph.
class TextRun {
  const TextRun(this.text, {this.bold = false});

  final String text;
  final bool bold;

  @override
  bool operator ==(Object other) =>
      other is TextRun && other.text == text && other.bold == bold;

  @override
  int get hashCode => Object.hash(text, bold);

  @override
  String toString() =>
      'TextRun("${text.replaceAll('\n', r'\n')}", bold: $bold)';
}

/// Parses [source] into paragraphs of [TextRun]s. See the library doc comment
/// for the grammar. Returns an empty list for blank input.
List<List<TextRun>> parseBoldMarkup(String source) {
  final paragraphs = source
      .split(RegExp(r'\n\n+'))
      .map((p) => p.trim())
      .where((p) => p.isNotEmpty);

  final result = <List<TextRun>>[];
  for (final para in paragraphs) {
    result.add(_runsForParagraph(para));
  }
  return result;
}

final _boldPattern = RegExp(r'\*\*([^*]+?)\*\*');

List<TextRun> _runsForParagraph(String para) {
  final runs = <TextRun>[];
  var cursor = 0;
  for (final m in _boldPattern.allMatches(para)) {
    if (m.start > cursor) {
      runs.add(TextRun(para.substring(cursor, m.start)));
    }
    runs.add(TextRun(m.group(1)!, bold: true));
    cursor = m.end;
  }
  if (cursor < para.length) {
    runs.add(TextRun(para.substring(cursor)));
  }
  // Drop empty plain runs (e.g. the "" before a leading match).
  runs.removeWhere((r) => !r.bold && r.text.isEmpty);
  return runs.isEmpty ? [const TextRun('')] : runs;
}

File: apps/web/src/lib/boldMarkup.ts
/**
 * A minimal inline-markup parser for Knowledge Note bodies.
 *
 * The grammar is deliberately tiny — NOT Markdown:
 * - `\n\n` (one or more blank lines) splits paragraphs; each paragraph is
 *   trimmed and empty paragraphs are dropped.
 * - `**x**` (x non-empty, no nested `**`, non-greedy) marks a bold run.
 * - Everything else is literal — a lone `**` renders as the two characters,
 *   and `*` / `_` / `#` are plain text. No nesting, italics, or links.
 *
 * Port of `lib/core/utils/bold_markup.dart` — keep in exact sync. The shared
 * test vectors (`test/fixtures/bold_markup_vectors.json`) guard both.
 */

/** One styled span of text within a paragraph. */
export interface TextRun {
  text: string;
  bold: boolean;
}

const BOLD_PATTERN = /\*\*([^*]+?)\*\*/g;

/**
 * Parses `source` into paragraphs of {@link TextRun}s. See the module doc
 * comment for the grammar. Returns an empty array for blank input.
 */
export function parseBoldMarkup(source: string): TextRun[][] {
  return source
    .split(/\n\n+/)
    .map((p) => p.trim())
    .filter((p) => p.length > 0)
    .map(runsForParagraph);
}

function runsForParagraph(para: string): TextRun[] {
  const runs: TextRun[] = [];
  let cursor = 0;
  BOLD_PATTERN.lastIndex = 0;
  let m: RegExpExecArray | null;
  while ((m = BOLD_PATTERN.exec(para)) !== null) {
    if (m.index > cursor) {
      runs.push({ text: para.slice(cursor, m.index), bold: false });
    }
    runs.push({ text: m[1], bold: true });
    cursor = m.index + m[0].length;
  }
  if (cursor < para.length) {
    runs.push({ text: para.slice(cursor), bold: false });
  }
  // Drop empty plain runs (e.g. the "" before a leading match).
  const nonEmpty = runs.filter((r) => r.bold || r.text.length > 0);
  return nonEmpty.length > 0 ? nonEmpty : [{ text: "", bold: false }];
}
